mcs/class/corlib/System/AppDomain.cs

```diff
@@ -1459,6 +1462,31 @@ internal byte[] GetMarshalledDomainObjRef ()
 
 		public event EventHandler<FirstChanceExceptionEventArgs> FirstChanceException;
 
+		static void FireProcessExit (object state)
+		{
+			try {
+				var procExit = CurrentDomain.ProcessExit;
+				if (procExit != null)
+					procExit (CurrentDomain, null);
+			} catch (Exception e) {
+				NonFatalUnhandledException (e);
+			} finally {
+				lock (state) {
+					Monitor.Pulse (state);
+				}
+			}
+		}
+
+		//XXX This is called by the runtime shutdown process, see runtime.c
+		static bool RunProcessExit ()
+		{
+			object obj = new object ();
+			lock (obj) {
```

Comment:
Is this guaranteed to be called only once?

Reply:
Yes, it's called once per domain by the runtime.